Feature: Add LocalFileMetadataSource and S3FileCopySink.

caskchester opened the pull request:
1. Added LocalFileMetadataSource that can read from local filesystem
   or local HDFS.
2. Added S3FileCopySink that can copy files into AmazonS3 filesystem.
3. Changed HDFSFileCopySink -> LocalFileCopySink. Now can copy files
   into Local HDFS or local filesystem.
4. Changed some Abstract classes to regular classes such that LocalFile
   plugins can directly use them without having to create more classes that
   extend from them.
5. Improved javadocs and comments.

Commit:
Feature: Add LocalFileMetadataSource and S3FileCopySink.
1. Added LocalFileMetadataSource that can read from local filesystem
   or local HDFS.
2. Added S3FileCopySink that can copy files into AmazonS3 filesystem.
3. Changed HDFSFileCopySink -> LocalFileCopySink. Now can copy files
   into Local HDFS or local filesystem.
4. Changed some Abstract classes to regular classes such that LocalFile
   plugins can directly use them without having to create more classes that
   extend from them.
5. Improved javadocs and comments.

anwar6953 commented on docs/LocalFileMetadataSource-batchsource.md:
```diff
@@ -0,0 +1,41 @@
+| **group**              | String | The group that the file belongs to.                                                                                                            |
+| **owner**              | String | The owner of the file.                                                                                                                         |
+| **isFolder**           | Boolean| Whether or not the file is a folder.                                                                                                           |
+| **relativePath**       | String | The relative path is constructed by deleting the portion of the source path that comes before the last path separator ("/") from the full path.|
```
How is this description different than just the fileName?

caskchester replied:
Here's an example:
Say you do `cp -r  /foo/bar   /somefolder/` And there's a file `/foo/bar/baz/bak.txt`
The sourcePath in this case would be `/foo/bar`.
The full path of the file is `/foo/bar/baz/bak.txt`
The filename is only `bak.txt`
The relative path, constructed by dropping the prefix, would be `bar/baz/bak.txt`
The copied file would have path `/somefolder/bar/baz/bak.txt`

You can find the parsing logic in FileMetadata.java

anwar6953 commented on docs/LocalFileMetadataSource-batchsource.md:
```diff
@@ -0,0 +1,41 @@
+| **modificationTime**   | long   | The modification timestamp of the file.                                                                                                        |
+| **group**              | String | The group that the file belongs to.                                                                                                            |
+| **owner**              | String | The owner of the file.                                                                                                                         |
+| **isFolder**           | Boolean| Whether or not the file is a folder.                                                                                                           |
```
I think directory is better-fitting terminology.

anwar6953 commented on docs/LocalFileMetadataSource-batchsource.md:
```diff
@@ -0,0 +1,41 @@
+| :--------------------- | :----- | :-------------------------                                                                                                                     |
+| **fileName**           | String | Only contains the name of the file.                                                                                                            |
+| **fullPath**           | String | Contains the full path of the file in the source file system.                                                                                  |
+| **fileSize**           | long   | File Szie in bytes.                                                                                                                            |
```
misspelling - size

anwar6953 commented on src/main/java/co/cask/hydrator/plugin/batch/file/hdfs/LocalFileCopySink.java:
```diff
@@ -0,0 +1,99 @@
+            throw new IllegalArgumentException(e.getMessage());
+          }
+          break;
+        case "hdfs" :
```
why just `break` if its "hdfs"? Based upon the name of this class (LocalFileCopyOutputFormatProvider), it doesn't seem like it should handle "hdfs"

caskchester replied:
This plugin only copies to local HDFS, hence the name "LocalFileCopySink".

The `FileSystem` instance created from a default `new Configuration()` points to the local HDFS. That's why no additional configurations are added.

anwar6953 replied:
LocalFileCopySink makes it seem like it uses the local file system (And not the hdfs on the cluster where its running).

caskchester replied:
Yeah I'll probably have to rename it. Any ideas? Or should I create 2 separate plugins?

anwar6953 replied:
Maybe just name it `FileCopySink`?

anwar6953 commented on widgets/S3FileCopySink-batchsink.json:
```diff
@@ -0,0 +1,93 @@
+        },
+        {
+          "widget-type": "select",
+          "label": "Amazon Region",
```
Is the amazon region set likely to grow?
If so, one way to handle that is to allow a textbox for user to enter in the region if it doesn't exist in this dropdown.

caskchester commented:
Addressed comments:
1. renamed LocalFilePlugins to FilePlugins.
2. Renamed isFolder to isDir
3. Removed configuration for S3 region. It isn't required for
   connecting to S3 filesystems and using the wrong region would
   cause the pipeline to fail. For more info look at the link below
   https://hadoop.apache.org/docs/r2.7.2/hadoop-aws/tools/hadoop-aws/index.html#S3A

Please review thanks!

Commit:
Addressed comments:
1. renamed LocalFilePlugins to FilePlugins.
2. Renamed isFolder to isDir
3. Added output schema in configurePipeline stage for source plugins.
4. Removed configuration for S3 region. It isn't required for
   connecting to S3 filesystems and using the wrong region would
   cause the pipeline to fail. For more info look at the link below
   https://hadoop.apache.org/docs/r2.7.2/hadoop-aws/tools/hadoop-aws/index.html#S3A

Commit:
killed unused imports

anwar6953 commented on src/main/java/co/cask/hydrator/plugin/batch/file/fs/FileCopySink.java:
```diff
@@ -0,0 +1,99 @@
+            conf.put(FileCopyOutputFormat.FS_HOST_URI,
+                     new URI(config.getScheme(), null, Path.SEPARATOR, null).toString());
+          } catch (URISyntaxException e) {
+            throw new IllegalArgumentException(e.getMessage());
```
pass in `e` instead of `e.getMessage()`, so you don't lose out on propagating the other information, such as the stacktrace.

anwar6953 replied:
Remainder LGTM

Commit:
changed e.getmessage to e
